libnetwork/drivers/bridge/interface_linux.go

```diff
@@ -53,39 +53,76 @@ func (i *bridgeInterface) exists() bool {
-func (i *bridgeInterface) addresses() ([]netlink.Addr, []netlink.Addr, error) {
+// addresses returns a bridge's addresses, IPv4 (with family=netlink.FAMILY_V4)
+// or IPv6 (family=netlink.FAMILY_V6).
+func (i *bridgeInterface) addresses(family int) ([]netlink.Addr, error) {
 	if !i.exists() {
 		// A nonexistent interface, by definition, cannot have any addresses.
-		return nil, nil, nil
+		return nil, nil
 	}
-	v4addr, err := i.nlh.AddrList(i.Link, netlink.FAMILY_V4)
+	addrs, err := i.nlh.AddrList(i.Link, family)
 	if err != nil {
-		return nil, nil, fmt.Errorf("Failed to retrieve V4 addresses: %v", err)
+		return nil, fmt.Errorf("Failed to retrieve addresses: %v", err)
 	}
+	return addrs, nil
+}
 
-	v6addr, err := i.nlh.AddrList(i.Link, netlink.FAMILY_V6)
-	if err != nil {
-		return nil, nil, fmt.Errorf("Failed to retrieve V6 addresses: %v", err)
+func (i *bridgeInterface) getRequiredIPv6Addrs(config *networkConfiguration) map[string]*net.IPNet {
+	requiredAddrs := map[string]*net.IPNet{}
+
+	// TODO(robmry) - is config.AddressIPv6 always set at this point?
+	//   The logic here is preserved from the original setupBridgeIPv6(), but
+	//   can probably be simplified.
+
+	// Store default link-local bridge network and default gateway, unless the user
+	// asked for a specific link local prefix.
+	if config.AddressIPv6 == nil || !config.AddressIPv6.IP.IsLinkLocalUnicast() {
+		i.bridgeIPv6 = bridgeIPv6
+		i.gatewayIPv6 = i.bridgeIPv6.IP
+		requiredAddrs[i.bridgeIPv6.String()] = i.bridgeIPv6
 	}
 
-	if len(v4addr) == 0 {
-		return nil, v6addr, nil
+	// Store the user-specified bridge network and gateway.
+	if config.AddressIPv6 != nil {
+		i.bridgeIPv6 = config.AddressIPv6
+		i.gatewayIPv6 = config.AddressIPv6.IP
+		requiredAddrs[i.bridgeIPv6.String()] = i.bridgeIPv6
 	}
-	return v4addr, v6addr, nil
+
+	return requiredAddrs
 }
 
-func (i *bridgeInterface) programIPv6Address() error {
-	_, nlAddressList, err := i.addresses()
+func (i *bridgeInterface) programIPv6Addresses(config *networkConfiguration) error {
+	// Get the IPv6 addresses currently assigned to the bridge, if any.
+	existingAddrs, err := i.addresses(netlink.FAMILY_V6)
 	if err != nil {
 		return &IPv6AddrAddError{IP: i.bridgeIPv6, Err: fmt.Errorf("failed to retrieve address list: %v", err)}
 	}
-	nlAddr := netlink.Addr{IPNet: i.bridgeIPv6}
-	if findIPv6Address(nlAddr, nlAddressList) {
-		return nil
+
+	// Get the list of required IPv6 addresses for this bridge, and assign values to
+	// "i.bridgeIPv6", "i.gatewayIPv6".
+	requiredAddrs := i.getRequiredIPv6Addrs(config)
+
+	// Remove addresses that aren't required, and remove already-existing required
+	// addresses from the map.
+	for _, existingAddr := range existingAddrs {
+		existingAddrStr := existingAddr.String()
+		_, required := requiredAddrs[existingAddrStr]
+		if required {
+			delete(requiredAddrs, existingAddrStr)
+		} else {
+			existingAddr := existingAddr
```

Comment:
Of course it's the gosec linter, which (intentionally?) errs on the side of false positives because security something something. (Alarm fatigue? Never heard of it 🙄) We have proven through static analysis that the callee does not hold a reference to the pointer, though I wish gosec could have done the escape analysis itself.